src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItemLoader.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Razor.Hosting
+{
+    /// <summary>
+    /// A loader implementation that can load <see cref="RazorCompiledItem"/> objects from an
+    /// <see cref="Assembly"/> using reflection.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// Inherit from <see cref="RazorCompiledItemLoader"/> to customize the behavior when loading 
+    /// <see cref="RazorCompiledItem"/> objects from an <see cref="Assembly"/>. The default implementations of methods
+    /// defined by this class use reflection in a trivial way to load attributes from the assembly.
+    /// </para>
+    /// <para>
+    /// Inheriting from <see cref="RazorCompiledItemLoader"/> is useful when an implementation needs to consider
+    /// additional configuration or data outside of the <see cref="Assembly"/> being loaded.
+    /// </para>
+    /// <para>
+    /// Subclasses of <see cref="RazorCompiledItemLoader"/> can return subclasses of <see cref="RazorCompiledItem"/>
+    /// with additional data members by overriding <see cref="CreateItem(RazorCompiledItemAttribute)"/>.
+    /// </para>
+    /// </remarks>
+    public class RazorCompiledItemLoader
+    {
+        /// <summary>
+        /// Loads a list of <see cref="RazorCompiledItem"/> objects from the provided <see cref="Assembly"/>.
+        /// </summary>
+        /// <param name="assembly">The assembly to search.</param>
+        /// <returns>A list of <see cref="RazorCompiledItem"/> objects.</returns>
+        public virtual IReadOnlyList<RazorCompiledItem> LoadItems(Assembly assembly)
+        {
+            if (assembly == null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            var items = new List<RazorCompiledItem>();
+            foreach (var attribute in LoadAttributes(assembly))
```

Comment:
nit: split the `LoadAttributes` into its own local variable for easier debugging.